%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.html and the instructions for   %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and 
%comment line below
%\documentclass{bmcart}

%%% Load packages
%\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
%\RequirePackage{hyperref}
\usepackage[utf8]{inputenc} %unicode support
\usepackage{listings}
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\usepackage{xr}
\externaldocument{chipPCR}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%\def\includegraphic{}
%\def\includegraphics{}
\usepackage{graphicx}
\graphicspath{ {./figures/} }


%%% Put your definitions there:
\startlocaldefs
\endlocaldefs

%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Software}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{chipPCR: an R Package to Pre-Process Amplification Curve Data}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Specify information, if available,       %%
%% in the form:                             %%
%%   <key>={<id1>,<id2>}                    %%
%%   <key>=                                 %%
%% Comment or delete the keys which are     %%
%% not used. Repeat \author command as much %%
%% as required.                             %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\author[
   addressref={aff1},                   % id's of addresses, e.g. {aff1,aff2}
   corref={aff1},                       % id of corresponding address, if any
   email={stefan.roediger@hs-lausitz.de}   % email address
]{\inits{S}\fnm{Stefan} \snm{R\"odiger}}
\author[
   addressref={aff2},
   email={michalburdukiewicz@gmail.com}
]{\inits{M}\fnm{Micha\l{}} \snm{Burdukiewicz}}
\author[
   addressref={aff1},
   email={peter.schierack@hs-lausitz.de}
]{\inits{P}\fnm{Peter} \snm{Schierack}}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%% Repeat \address commands as much as      %%
%% required.                                %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address[id=aff1]{%                           % unique id
  \orgname{Faculty of Natural Sciences, Brandenburg University of Technology 
Cottbus--Senftenberg}, % university, etc
  \street{Gro\ss{}enhainer 57},                     %
  \postcode{01968}                                % post or zip code
  \city{Senftenberg},                              % city
  \cny{Germany}                                    % country
}
\address[id=aff2]{%
  \orgname{Department of Genomics, Faculty of Biotechnology, University of 
Wroc\l{}aw},
  \street{ul. Fryderyka Joliot-Curie 14a},
  \postcode{50-383}
  \city{Wroc\l{}aw},
  \cny{Poland}
}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter short notes here                   %%
%%                                          %%
%% Short notes will be after addresses      %%
%% on first page.                           %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{artnotes}
%\note{Sample of title note}     % note to the article
%\note[id=n1]{Equal contributor} % note, connected to author
\end{artnotes}

%\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{abstractbox}

\begin{abstract} % abstract
\parttitle{Background}
The quantitative real-time polymerase chain reaction (qPCR) and quantitative 
isothermal amplification (qIA) are standard methods for quantification of 
nucleic acids. Numerous real-time read-out technologies with different technical 
foundation have been developed. Despite the continuous interest in amplification 
based techniques, there are only few tools for amplification data 
pre-processing. It is a major setback especially during development of new 
instruments, when the precise control on raw data is indispensable.

\parttitle{Results} $\emph{chipPCR}$ is an \textbf{R} package for pre-processing 
and quality analysis of amplification curve data from qPCR and qIA experiments. 
The package takes advantage of \textbf{R}â€™s \emph{S4} object model and offers an 
extensible environment. Functions to simulate, normalize, baseline, impute 
missing values, to smooth amplification curves and a function to to detect the 
start and end of an amplification curve are part of the software. We implemented 
a 5-point stencil for derivative interpolation, which is unique in the 
\textbf{R} landscape. In additions, offers $\emph{chipPCR}$ statistical tests 
and plots for data quality management and efficiency calculation. Data sets 
encompass experiments from helicase dependent amplification, capillary 
convective PCR and qPCR under various temperature conditions and detection 
systems (e.g., hydrolysis probes, intercalating dyes). Functions were 
implemented as GUI desktop or server applications using $\emph{shiny}$ 
technology for report generation. The structure of the packages is open for 
integration to Web based and standalone \emph{shiny} applications. 

\parttitle{Availability} http://cran.r-project.org/web/packages/chipPCR
\parttitle{Contact} stefan.roediger@hs-lausitz.de
\parttitle{Supplementary information} Supplementary data are available at
Bioinformatics online.
\end{abstract}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The keywords begin here                  %%
%%                                          %%
%% Put each keyword in separate \kwd{}.     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{keyword}
\kwd{R package}
\kwd{Amplification curve}
\kwd{Data analysis}
\kwd{Quantitative polymerase chain reaction}
\kwd{Quantitative isothermal amplification}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
\end{fmbox}% uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%% start of article main body
% <put your article body there>

%%%%%%%%%%%%%%%%
%% Background %%
%%
\section*{Background}

Quantitative polymerase chain reaction (qPCR) and quantitative isothermal 
amplification (qIA) are standard methods to amplify nucleic acids. Recently, 
continuous temperature gradient technologies emerged (e.g., microfluidics) 
\cite{rodiger_nucleic_2014}.These methods are used in real-time monitoring 
technologies, such as our previously reported VideoScan technology, microfluidic 
systems, point-of-care devices, and conventional real-time thermo cyclers 
\cite{rodiger_nucleic_2014,chang_2012,roediger_highly_2013}. Real-time 
technologies enable the quantification of nucleic acids by calculation of 
specific curve parameters like the quantification point (Cq) and the 
amplification efficiency (AE) \cite{ruijter_2013,pabinger_2014}. Fundamental 
steps of amplification curve analysis are: (1) raw data read-in, (2) 
amplification curve pre-processing (e.g., noise reduction, outlier removal), (3) 
amplification curve processing (e.g., Cq and AE calculation), (4) 
post-processing and quantification of secondary parameters (e.g., Delta-Delta-Ct 
for gene expression analysis), (5) data export, (6) visualization and (7) report 
generation.

  \textbf{R} belongs to the most used bioinformatics tools and is known to be an 
early adopter of emerging technologies such as digital PCR and  NanoString 
nCounter Platform \cite{waggott_2012,pabinger_2014}. At same time it is widely 
used for established methods, such as qPCR. Most \textbf{R} packages focus on 
the read-in and (post)-processing of data sets, which originate from commercial 
qPCR systems. Sophisticated \textbf{R} packages for the steps 1. and 3.--7. are 
available from Bioconductor and CRAN 
\cite{pabinger_2014,Dvinge_2009,perkins_2012,heckmann_2011,Zhang_2010, 
mccall_2014,gehlenborg_2013,huntley_2013,zhang_2013}. However, 
there is no \textbf{R} package for pre-processing and quality analysis of 
amplification curve data. Similar applies to other software solutions (compare 
\cite{pabinger_2014}). Pre-processing is in most commercial cyclers a ``black 
box''. This has limitations, such as hard to reproduce analysis on other 
platforms, difficulties during the adoption to changing experimental setups and 
limited sophisticated statistical tools. Moreover, it is desirable set up 
workflows in an open environment, which enables downstream analyses and which 
offers powerful tools for data visualizations and automatic report generation. 
Novel developed qPCR and qIA technologies initially depend on tools to 
pre-process the raw data. Pre-processing specifically addresses raw data 
inspection, steps to transform raw data in a compatible format for successive 
analysis steps (e.g., smoothing, imputation of missing values), data reduction 
(e.g., removal of invalid sets) and data quality management. The data quality of 
experimental instruments is often not ready for end-user analysis and 
presentation. It is important to use as many raw data as possible. 
Pre-processing algorithms remove stochastic errors and artefacts (e.g., noise, 
photo-bleaching effects, degassing effects, different signal levels) illustrated 
in Figure~\ref{figure:problems}. Misinterpretations are more likely if non or 
manual (``arbitrary'') corrections are performed. A manual alteration is in 
contradiction to reproducible research. The $\emph{chipPCR}$ package 
($\emph{chipPCR}$: ``Lab-on-a-\textbf{Chip}'' \& \textbf{PCR}) was developed to 
automatize pre-processing, to ease data analysis/visualization and to offer a 
quality control for the statistical data analysis of qPCR and qIA experiments 
and is primarily targeted at developers but non the less at users who process 
raw data of commercial systems. 


\section*{Implementation}

$\emph{chipPCR}$ package was implemented in the \textbf{R} software environment 
for statistical computing and graphics according to the guidelines as described 
elsewhere \cite{RDCT2014a}. The supplemental material, available as the packages 
vignettes, use Donald Knuth's literate programming principle~\cite{Knuth1984} to 
present the source code in the most convenient way for the users. The package is 
available at CRAN 
(http://cran.r-project.org/web/packages/chipPCR/index.html) along with an 
extensive documentation and examples. The latest source code is hosted at GitHub 
(https://github.com/michbur/chipPCR) with an issue tracker for bug reports 
or feature requests. For the present paper $\emph{chipPCR}$ version~0.0.8-2 was 
used. The main naming convention in $\emph{chipPCR}$ is 
\textit{period.separated} names \cite{Baaaath_2012}. $\emph{chipPCR}$ is a 
relative of \emph{MBmca} \cite{roediger_RJ_2013} and \emph{dpcR} 
\cite{burdukiewicz_2014} but is focused on the processing of amplification curve 
data.
  
  $\emph{chipPCR}$ uses \textbf{R}â€™s object model, designated \emph{S4} class 
system, which is implemented in the \emph{methods} package. In contrast to 
\textbf{R}â€™s \emph{S3} class system it is a requirement in \emph{S4} to declare 
classes, slots, methods relationships explicitly and to establish formal 
declarations of methods. This means that the number and types of slots in an 
instance of a class have to be established at the time the class definition and 
the objects from the class are validated against this definition and have to 
comply to it at any time. \emph{S4} methods are declared by calls to 
\textsl{setMethod} together with the names and signatures of the arguments. 
Signatures are used for identification of classes of one or more arguments of 
the methods. \textsl{setGeneric} can be used to declare generic functions. One 
disadvantage of \emph{S4} classes is a higher ``package author effort'' in 
comparison to \emph{S3} classes. However, \emph{S4} provides better assurance 
that each object in a class has the required slots, that data in the slots have 
consistent names and classes, and the possibility to include additional 
information (e.g., results, parameters) \cite{Karatzoglou_2004}.

Selected functionality is used in the $\emph{RDML}$ \cite{blagodatskikh_2014} packages.

Functions of the $\emph{RDML}$ package are listed in supplement Section~\ref{sec:functions}.

During the past years, several high-throughput technologies emerged, which 
require higher computing power. The core structures of $\emph{chipPCR}$ avoid 
loops to keep the code fast. Structures for parallel computing have been 
implemented in \textsl{smoother}. $\emph{chipPCR}$ includes a set of classes 
for plotting. The \textbf{R} graphics system allows fine control over the small 
details of plots \cite{Murrell_2012}. Although, we keep the output 
of our custom made plots simple, many parameters can be adjusted directly or by 
the ellipse parameter.

 $\emph{chipPCR}$ has no build-in functionality for data import or specialized 
parsers since \textbf{R} has numerous tools and auxiliary packages for 
customized parsers. Rather we have chose to rely on \textbf{R} workspaces, and 
dedicated \textbf{R} packages as default data format and standard import and 
export as described elsewhere \cite{perkins_2012, RDCT2014c, 
blagodatskikh_2014}. The results of $\emph{chipPCR}$ functions are usually \emph{S4} 
objects accompanied with specifically tailored summary and plot methods.

To spread the use among novice and less routinized users we believe that a 
graphical user interface (GUI) is important. Currently there are several 
\textbf{R} GUI projects \cite{Valero_2012} to chose from. Some functionality of 
$\emph{chipPCR}$ (e.g., removal of missing values, smoothing, ...) originates 
from experimental \textbf{RKWard} plugins \cite{roediger_bead_qPCR_2013, 
roediger_RKWard_2013}. \textbf{RKWard} \cite{rodiger_rkward_2012} is a 
cross-platform GUI and integrated development environment for \textbf{R}. 
However, the functionality of $\emph{chipPCR}$ went beyond the functions of the 
plugins. Recently, \emph{shiny} emerged as a very interesting framework to build 
GUIs for the desktop (web browser) or services for interactive web applications, 
which can be deployed virtually anywhere. \emph{shiny} is a technology to build 
web applications with highly customizable widgets (e.g., sliders, plots, tables, 
reports) in a plugin-like approach. The combination \textbf{R} and \emph{shiny} 
technology allows a simple and repaid extension of the software. The 
\emph{shiny} applications update live and interactively. The user interfaces can 
be built entirely using R and operates in any R environment (see \emph{shiny} 
manual for details). Currently, \emph{shiny} GUIs for the functions 
\textsl{AmpSim}, \textsl{bg.max} (\textsl{AmpSim.gui} app) and 
\textsl{amptester} (\textsl{amptester.gui} app) are available. However, it is 
possible to run the applications as service on a server (e.g., 
http://michbur.shinyapps.io/MFIaggr\_gui/), on the local desktop (e.g., 
Figure~\ref{figure:browser}), or as deployed from an external source for a local 
\textbf{R} installation 
(supplement Section~\ref{sec:AmpSim}). Similarly to 
\textbf{QPCR} \cite{pabinger_2009} is \emph{shiny} in combination with 
\textbf{R} the foundation to build monolithic systems to parse, pre-process and 
analyze amplification curve data in a combined work-flow. For example, the GUI 
embedding \textsl{MFIaggr.gui} $\emph{chipPCR}$ function is accessible as the 
web server service under the address:
http://michbur.shinyapps.io/MFIaggr\_gui/.

Data sets are an essential element of \textbf{R} packages \cite{hofmann_2013} 
and an indispensable requirement for reproducible research \cite{Leeper_2014}. 
The $\emph{chipPCR}$ package contains 22 data sets along with a description of 
the experimental procedure. The data sets stem from commercial qPCR cyclers 
(e.g., Bio-Rad iQ5, Bio-Rad CFX 96) and experimental systems. The later 
encompass the VideoScan heating/cooling unit (VideoScan HCU) 
\cite{roediger_highly_2013} and the capillary convective PCR (ccPCR) (see 
manual 
of $\emph{chipPCR}$). The experiments were performed under different condition 
including conventional qPCR, isothermal amplification by helicase dependent 
amplification (HDA) \cite{rodiger_nucleic_2014}.

\section*{Result and discussion}

$\emph{chipPCR}$ offers routines for pre-processing and quality management of 
experiments curves from, ccPCR, qIA and qPCR experiments (see supplement 
Section~\ref{sec:HDA}, \ref{sec:ccPCR}). Implemented as \textbf{R} console tool 
it can be accessed and operated by an integrated development environment, 
RScript or related approaches. The \emph{shiny} technology makes it applicable 
as web browser from network connected computers or as local service. This allows 
to analyze, pre-process and to visualize results.

\subsection*{Multiple comparison of the cycle dependent variance of the 
amplification curves}

\textsl{MFIaggr} is a powerful analytical and graphical tool for fast multiple 
comparison of the cycle dependent signal dispersion and distribution. The 
continuous response variable $y'$ is used to describe the relationships to one 
or more continuous predictor variables $y_1, ..., y_n$. Use cases include the 
comparison of independent reaction vessels or the analysis of replicate 
experiments (supplement Section~\ref{sec:MFIaggr}). In particular, this function 
might be useful for quality management during the development of high-throughput 
technologies. An analysis via a the \emph{shiny} \textsl{MFIaggr.gui} app is 
shown in Figure~\ref{figure:MFIaggr2}.

\subsection*{humanrater - a human machine interface}

The \emph{chipPCR} package has numerous function to build streamlined work-flows 
for unsupervised pre-processing (see subsequent sections). However, as any 
unsupervised system is prone to errors we designed \textsl{humanrater}, which is 
a graphical interface to rate (``qualitative'' and/or ``quantitative'') curves 
interactively in an \textbf{R} GUI. All curves can be analyzed in random 
sequence, to reduce the risk of bias, or natural sequence. The number of 
replicate rating runs is user definable. A curve and prompts an input field for 
the user. This function can be used to the human rating and the rating of a 
machine. Results of the analysis can be used for further processing 
(see supplement Section~\ref{sec:humanrater}).

\subsection*{Imputation of missing values in amplification curves}

  Experimental technologies may produce missing values (NA) at 
random due to sensor drop-outs or other technical difficulties. Many analytical 
functions stop to progress or discard entire data sets. This 
behavior is rational for unknown data structures. However, in case of 
amplification curve data it is justified to impute NAs because the structure 
generally resembles an S-shaped curve. Standard approaches include substitution 
with most frequent values, mean value imputation, last value carried forward, 
bootstrapping, or substitution by correlation with replicate measurements 
\cite{Harrell_2001}. In case of amplification curves other approaches are 
favorable. Particularly, the transitions phases (e.g., background phase to 
exponential phase) is potentially prone to bias.

  The function \textsl{fixNA} imputes missing values in a single column 
of data (response). The imputation is based either on a linear approximation or an 
approximation by cubic splines (default) (Figure~\ref{figure:fixNA}). 
Experiments showed that cubic spline interpolation yielded the most probable 
values and therefore led to the least effect on tested statistical parameters 
(Cq, background signal, Pearson correlation coefficient) the exponential phase 
and is therefore the recommended approach to remove missing values 
(Figure~\ref{figure:fixNA_data}). We observed no significant bias by cubic spline 
interpolation (Table~\ref{table:fixNA}). The performance of \textsl{fixNA} using 
cubic splines was better than a linear interpolation (Figure~\ref{figure:fixNA}). 
However, the linear approximation might be applicable in measurements with high 
sample rates (e.g., isothermal amplification) (not shown). Any method requires a 
minimum number of data points as foundation for a meaningful imputation. 
\textsl{fixNA} attampts to take care of such pitfalls. By rule of thumbs we 
determined that the number of missing elements in relation to the total number 
of elements. In case more than 30~\% of all values are NAs gives \textsl{fixNA} 
a warning.

\subsection*{Smoothing and filtering}

Amplification curve data of experimental thermo-cyclers may deliver results, 
which are hard to interpret due to noise. For data presentation it is often 
useful to smooth or filter the data. Therefore, we developed the 
\textsl{smoother} function, which is a wrapper for smoother functions and 
filters commonly used to process amplification curve data (supplement 
Section~\ref{sec:smoothing_filterring}). \textsl{smoother} inherited traits 
(supplement Table~\ref{table:tab1})  of the parent functions. However, the 
functionality of \textsl{smoother} greatly outgrowths applications only in 
amplification curve analysis. Incorporating most of the best proven algorithms, 
we offer the user a powerful tool to access the methods while minimizing the 
drawback of learning syntax of specific functions. \textsl{smoother} was 
enhanced by functionality of \textsl{fixNA} and \textsl{CPP}. 
Figure~\ref{figure:smoother} shows results of the \textsl{smoother} function an 
amplification curve data. Many functions (e.g., Savitsky-Golay filter) of 
$\emph{chipPCR}$ assume uniform (equally spaced) sampling. The function 
\textsl{smoother} and \textsl{CPP} (inherited from \textsl{smoother}) give a 
warning in such cases. The \textsl{smoother} function enables users to tune 
behavior of the chosen smoothing algorithm by using nearly all parameters 
available in called subroutines and at the same time uniforms input and output.

\subsection*{Normalization of amplification curve data}

It is a common characteristic of amplification curve data that the fluorescence 
values between samples vary due to high background, sample inhomogeneities and 
variances in dye quantities (Figure~\ref{figure:normalization}~\emph{A}). 
Therefore, normalization of amplification curve data is a common task during the 
data analysis. To scale the fluorescence between 0 and 1 a \emph{Min-Max 
normalization} (Equation~\ref{eq:normalization}) can be used 
\cite{roediger_RJ_2013}. In addition, we propose an alternative normalization based on 
quantiles (Equation~\ref{eq:quantile_normalization}). We argue that quantiles are less 
affected by outliers. In addition a normalization to maximum 
(Equation~\ref{eq:max_normalization}, 
Figure~\ref{figure:normalization}~\emph{D}) and by standard score 
(Equation~\ref{eq:zscore_normalization}, 
Figure~\ref{figure:normalization}~\emph{F}).

  The slope in a curve can be corrected by a linear regression. \textsl{lm.coefs} 
offers four linear regression models to calculate the slope based on the 
background range. This includes a ordinary least squares method (\textsl{lm}, 
\emph{stats}) but also three robust methods. The robust regression methods are 
considered to be less vulnerable to outliers. This feature is especially useful, 
when the background range contains considerable noise. The methods are (I) a 
nonparametric rank-based estimator \cite{Kloke_2012}, (II) quantile regression 
\cite{Koenker_2008} and (III) a MM-type estimators for linear regression 
\cite{Todorov_2009}. 

\subsection*{Interpolation of derivatives}
  Many methods for curves analysis require the calculation of derivatives. It is 
possible to solve this by fitting a curve to a function and performing symbolic 
derivation. Unfortunately, this approach causes information loss through the fit 
and unnecessary adds additional assumptions regarding the relation between cycle 
number and fluorescence level. Hence, we integrated the \textsl{inder} function. 
\textsl{inder} (``in'' and ``der'' = interpolate derivatives) finds numeric 
derivatives by a five-point stencil, a commonly used finite difference method 
(supplement Section~\ref{sec:inder}). These methods approximate derivative in a 
given point by adding up products of nearby values of function and their 
weights~\cite{Dahlquist_2008}. This function can be used to estimate the 
approximate cycle of quantification (Cq). Differentiation is a method for 
background suppression and reduction of the inter sample background amplitude 
variations (Figure~\ref{figure:inder_fit}~A~and~B). Smoothing may enhance the 
calculation of derivatives calculation and optimize the signal-to-noise ratio. 
Therefore, we implemented spline interpolation.

\subsection*{Estimating the start and the end of amplification process}

  The correction of background fluorescence is an important steps in 
amplification curve analysis. Background herein refers to a level of 
fluorescence measured before any specific amplification is detectable. It has 
influence on amplification efficiency calculation and model fitting 
\cite{tuomi_2010, rutledge_2008, ruijter_2009}. There are numerous ways of 
compensating background noise, ranging from simple to very sophisticated 
solutions (see supplement for details). The function 
\textsl{bg.max} is a fit-free method to estimate the range of the background 
starting from the raw data. The raw data (e.g., fluorescence intensity) measured 
after each step (cycle or time point) follow a non-linear progress. The 
\textsl{bg.max} algorithm assumes that the signal difference $\Delta{y}$ of 
successive cycles ($\Delta{y} = y_{n + 1} - y_n$) in the linear ground phase is 
approximately constant. After transition in the early exponential phase the 
signal changes drastically. First data are smoothed by Friedmanâ€™s â€™super 
smootherâ€™ (\textsl{supsmu}, \emph{stats}). Thereof, the approximate first and 
second derivative are calculated. The difference of cycles at the maxima of the 
first and second approximate derivative and a correction factor are used to 
estimate the range before the exponential phase. \textsl{bg.max} finds the 
background range without modeling of a fit function. The start of the background 
range is defined be a fixed value. Since many signals tend to overshot in the 
first cycles a default value of two is chosen. \textsl{bg.max} attampts to 
estimate the end of an amplification reaction (Figure~\ref{figure:bgmax}, 
supplement Section~\ref{sec:bgmax}).

\subsection*{Proposed work-flow}

The function \textsl{CPP} (``Curve Pre-Processor'') is a wrapper for functions 
described before in this study. \textsl{CPP} can be considered as proposed 
work-flow for an amplification curve pre-processing 
(Figure~\ref{figure:workflow}). In particular, the functions \textsl{fixNA} 
$\rightarrow$ \textsl{bg.max} $\rightarrow$ \textsl{smoother} $\rightarrow$ 
\textsl{normalizer} $\rightarrow$ \textsl{amptester} were cascaded. A typical 
output of \textsl{CPP} is shown in Figure~\ref{figure:HDA}, 
\ref{figure:thcyc_ccPCR}, \ref{figure:CPP_C54} and \ref{figure:normalization}.

\section*{Conclusions}

To the best of our knowledge is the $\emph{chipPCR}$ is the first \textbf{R} 
package for the pre-processing and raw data quality analysis of amplification 
curve data. Others and we share the philosophy that software in research should 
not be a black-box and but part of an transparent ecosystem for reproducible 
research and exploration of new information from existing data 
\cite{roediger_RJ_2013, hofmann_2013, Leeper_2014,Thioulouse_2010,liu_2014}. 
In particular, the smoothing methods and algorithms for the derivatives are 
often not documented and therefore hard to judge by others. The $\emph{chipPCR}$ 
package is open source (with GPL-3 license) and freely available through CRAN. 
$\emph{chipPCR}$ primarily targets pre-processing. However, the $\emph{chipPCR}$ 
has also implementations to process amplification curve data. These can be 
embedded in customized routines with other packages mentioned in this study 
(see supplement). In fact, the packages \emph{dpcR} and \emph{MBmca} depend on 
$\emph{chipPCR}$ technology. We showed that $\emph{chipPCR}$ is build from 
smaller blocks and show how users can do estimation of background by hand, 
solely by \textsl{inder}, \textsl{smoother} (\textsl{smoother} will be a method 
of smoothing in \textsl{inder}) and by putting data in \textsl{bg} object with 
\textsl{summary-der} for $SDm$ and $SDM$. It should be quite easy even for an 
inexperienced user. We can claim that modular structure of $\emph{chipPCR}$ 
package allows user to perform flexible data analysis adjusted to their needs. A 
limitation of all \textbf{R} packages related to qPCR and qIA is the lack of a 
comprehensive GUI. We argue that that our \emph{shiny} GUI approach might be a 
promising approach. Similar applies to a standard for data exchange. A 
standardized might be embedded in structures for models (e.g., Predictive Model 
Markup Language (PMML) \cite{Guazzelli_2009}) and data exchange (e.g., XML-based 
Real-Time PCR Data Markup Language (RDML) 
\cite{blagodatskikh_2014,lefever_2009}, binary formats \cite{michna_2013}). 
However, work on this hasn't started yet. In high-throughput systems might 
computing speed be an issue. Further parallelization of 
processes as described elsewhere \cite{Schmidberger_2009, boehringer_2013} is 
part of future developments. 

\section*{Availability and requirements}
Project name: chipPCR, 
Project homepage (development):
https://github.com/michbur/chipPCR, 
Project homepage at CRAN: 
http://cran.r-project.org/web/packages/chipPCR/index.html, 
Operating System: Platform independent, 
Other requirement: R 3.1.0 or higher, 
License: GPL-3

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Competing interests}
The authors declare that they have no competing interests.

\section*{Authors' contributions}
SR conceived of the study, and participated in its design and coordination and 
wrote the manuscript. SR and MB jointly developed the software. All authors read 
and approved the final manuscript. PS improved the manuscript.

\section*{Acknowledgment}
Part of this work was funded by the BMBF InnoProfile-Projekt 03 IPT 611X. 
Grateful thanks belong to all authors of the cited \textbf{R} packages, the 
\textbf{R} community and \textbf{RKWard} developers. We would like to thank 
Claudia Deutschmann (Brandenburg University of Technology Cottbus - Senftenberg, 
Germany) for the contribution of data.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file
\bibliography{roediger-burdukiewicz}      % Bibliography file (usually '*.bib' )

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

      
\begin{figure*}[h]
  \includegraphics[clip=true,trim=0cm 0cm 0cm 0cm,width=16cm]{MFIaggr}
  \caption{\label{figure:MFIaggr}
  \csentence{Amplification curve analysis of the \textsl{VIMCFX96\_60} data 
set.} The cycles 1 -- 10 were selected as region of interest (ROI). The 
function 
\textsl{MFIaggr} was used to analyze the variance of a 96-well plate cycler 
(Bio-Rad CFX96) and EvaGreen\textregistered for detection (left panel). The 
mean 
and the median 
were almost identical. The density plot (right upper panel) and 
quantile-quantile analysis (right lower panel). Both analytical plots indicate 
that the data of the background range are normal distributed. Note that both 
the 
$median \pm mad$ and the $mean \pm sd$ have similar results. }
      \end{figure*}       
  
     
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
% %%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{Additional Files}
  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.



\end{backmatter}
\end{document}
